### PR TITLE
Ясакова Татьяна. Задача 3. Вариант 14. Быстрая сортировка с простым слиянием.

### PR DESCRIPTION
**Pull request by Tanya-Yasakova**
Алгоритм включает в себя следующие основные этапы:
1. Декомпозиция данных:
    Данные равномерно распределяются между процессами с помощью scatterv.
    Каждый процесс сортирует свою часть данных с помощью итеративной быстрой сортировки.

 2. Взаимодействие между рабочими процессами и мастером:
    Рабочие процессы отправляют свои отсортированные элементы по одному мастеру.
    Перед отправкой каждого элемента рабочий процесс ждет подтверждения от мастера.
    Мастер-процесс использует минимальную кучу для эффективного слияния данных, поступающих от рабочих процессов.
    На каждом шаге мастер извлекает наименьший элемент из кучи, добавляет его в результирующий массив и запрашивает следующий элемент у соответствующего рабочего процесса.

3.  Синхронизация и сбор результатов:
    Функция merge_func вызывает master_function для мастера и worker_function для рабочих процессов.
    Все процессы синхронизируются через барьер, чтобы гарантировать завершение обмена данными.

Таким образом, алгоритм сочетает в себе декомпозицию данных, локальную сортировку и эффективное слияние результатов с помощью минимальной кучи, обеспечивая параллельную обработку и синхронизацию между процессами.

### CODE REVIEW

File: tasks/seq/yasakova_t_quick_sort_with_simple_merge/perf_tests/main.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
+#include <gtest/gtest.h>
+
+#include <algorithm>
+#include <random>
```
Comment (emilien-gus) [inline tasks/seq/yasakova_t_quick_sort_with_simple_merge/perf_tests/main.cpp:4]: Please, remove all unnecessary headers

File: tasks/mpi/yasakova_t_quick_sort_with_simple_merge/func_tests/main.cpp [hunk 2]
```diff
@@ -0,0 +1,179 @@
+#include <gtest/gtest.h>
+
```
Comment (emilien-gus) [inline tasks/mpi/yasakova_t_quick_sort_with_simple_merge/func_tests/main.cpp:2]: Please, add some tests with sorted data and with big numbers

File: tasks/mpi/yasakova_t_quick_sort_with_simple_merge/func_tests/main.cpp [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+#include <boost/mpi/environment.hpp>
+#include <random>
+#include <vector>
+
+#include "boost/mpi/collectives/broadcast.hpp"
+#include "core/task/include/task.hpp"
+#include "mpi/yasakova_t_quick_sort_with_simple_merge/include/ops_mpi.hpp"
+
+std::vector<int> create_random_integer_vector(int size, int minimum_value = -100, int maximum_value = 100,
+                                              unsigned random_seed = std::random_device{}()) {
```
Comment (AnastasiaRezantseva) [inline tasks/mpi/yasakova_t_quick_sort_with_simple_merge/func_tests/main.cpp:14]: move methods to your namespace

File: tasks/seq/yasakova_t_quick_sort_with_simple_merge/func_tests/main.cpp [hunk 4]
```diff
@@ -0,0 +1,109 @@
+#include <gtest/gtest.h>
+
+#include <algorithm>
+#include <vector>
+
+#include "seq/yasakova_t_quick_sort_with_simple_merge/include/ops_seq.hpp"
+
+void executeQuickSortTest(const std::vector<int>& inputData) {
+  std::vector<int> unsortedData = inputData;
+  std::vector<int> sortedData;
```
Comment (AnastasiaRezantseva) [inline tasks/seq/yasakova_t_quick_sort_with_simple_merge/func_tests/main.cpp:10]: move all methods to your namespace

File: tasks/mpi/yasakova_t_quick_sort_with_simple_merge/func_tests/main.cpp [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+TEST(yasakova_t_quick_sort_with_simple_merge_mpi, test_positive_and_negative_numbers_array) {
+  execute_parallel_sort_test({-5, -10, 5, 10, 0});
+}
+
+TEST(yasakova_t_quick_sort_with_simple_merge_mpi, test_large_random_numbers_array) {
+  execute_parallel_sort_test({123, 456, 789, 321, 654, 987});
+}
+
+TEST(yasakova_t_quick_sort_with_simple_merge_mpi, test_random_numbers_array) {
```
Comment (AnastasiaRezantseva) [inline tasks/mpi/yasakova_t_quick_sort_with_simple_merge/func_tests/main.cpp:140]: add tests with negative numbers

**Review by AnastasiaRezantseva — CHANGES_REQUESTED** [review]

**Review by AnastasiaRezantseva — APPROVED** [review]

**Review by emilien-gus — APPROVED** [review]

File: tasks/mpi/yasakova_t_quick_sort_with_simple_merge/src/ops_mpi.cpp [hunk 6]
```diff
@@ -0,0 +1,159 @@
+#include "mpi/yasakova_t_quick_sort_with_simple_merge/include/ops_mpi.hpp"
+
+#include <boost/serialization/array.hpp>
+#include <boost/serialization/vector.hpp>
+#include <queue>
+#include <stack>
+
+namespace yasakova_t_quick_sort_with_simple_merge_mpi {
+
+struct PriorityQueueNode {
```
Comment (allnes) [inline tasks/mpi/yasakova_t_quick_sort_with_simple_merge/src/ops_mpi.cpp:10]: please use std::priority_queue

Comment (Tanya-Yasakova) [reply tasks/mpi/yasakova_t_quick_sort_with_simple_merge/src/ops_mpi.cpp]: I already use std::priority_queue in master_function

Comment (allnes) [reply tasks/mpi/yasakova_t_quick_sort_with_simple_merge/src/ops_mpi.cpp:10]: please remove unused structure

Comment (Tanya-Yasakova) [reply tasks/mpi/yasakova_t_quick_sort_with_simple_merge/src/ops_mpi.cpp:10]: I use this: std::priority_queue<PriorityQueueNode, std::vector<PriorityQueueNode>

**Review by allnes — APPROVED** [review]
